src/config.cpp

```diff
@@ -302,6 +289,30 @@ Node Config::normalize_window(const Node &win,  // in: window to normalize
     return norm_win;
 }
 
+static unsigned int imx8_freq_i_from_freq(bool is_a53, size_t freq_MHz)
```

Comment:
I'm not happy that the information about available frequencies needs to be duplicated here. I wonder whether it's necessary to have this function and `Config::validate_per_process_freq_feasibility`. Wouldn't it be simpler to validate the feasibility later, e.g. in `DemosSched` constructor or `setup()` or at least in `config.create_scheduler_objects()`. At this points the `PowerPolicy` is already available and checking of power policy feasibility can be done better. Perhaps, the `PowerPolicy` class can have `validate()` method which would implement all the checks. Would there be any problem with that?

Reply:
Overall, the reason I implemented the checks in `Config` is to allow to user to statically validate the config without executing it. I agree it's not nice in the code, but the UX is better this way. :) However, it only works while we support only i.MX8.

The conversion from frequency in MHz to the frequency index is not necessary, but it's easier to work with, as I can use a bitmap for the AND/OR operations during the validation. It can be done using `std::set`, but it's a lot less efficient and the API is not very nice. I could create the bitmap dynamically before knowing the frequencies by creating a `std::map` `frequency => bitmap index` dynamically as I encounter the frequencies in config. That way `Config` doesn't need to know the exact frequencies and there's no duplication.

However, I still need to know the layout of the CPU clusters, which I cannot get outside `PowerPolicy`, and it depends on the current machine. Either I'll duplicate it in `Config` and it will be i.MX8-specific, or I can move the validation to `PowerPolicy` and make it dependent on the current machine. As I noted in the `Config::validate_per_process_freq_feasibility` docstring, it is possible to verify the feasibility later, and it would work for any CPU (not just i.MX8). However, the check wouldn't run during config parsing.

To move the validation to the power policy, it would need to know about either the `Config` instance, or all `Process` and `Window` instances. It is imo OK (design-wise) to pass them to a method like `void PowerPolicy::validate_configuration(Config c)`.


TL;DR: The validation can be kept in `Config` and improved to not depend on knowing the frequencies ahead-of-time, or it can be moved to PowerPolicy, where the check occurs later, depends on current machine, but it's more flexible and works everywhere (not just i.MX8). The first one is less work, the second one is probably a bit cleaner.